proxy/grpcproxy/cache/store.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2016 The etcd Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cache
+
+import (
+	"hash/adler32"
+	"sync"
+
+	"fmt"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/mvcc"
+	"github.com/golang/groupcache/lru"
+)
+
+const DefaultMaxEntries = 2048
+
+type Cache interface {
+	Add(req *pb.RangeRequest, resp *pb.RangeResponse)
+	Get(req *pb.RangeRequest) (*pb.RangeResponse, bool, error)
+	Compact(revision int64)
+}
+
+// keyFunc returns the key of an request, which is used to look up in the cache for it's caching response.
+func keyFunc(req *pb.RangeRequest) uint64 {
```

Comment:
let's do this instead

``` go
keyFunc(req *pb.RangeRequest) string {
    // todo: use marshalTo to reduce allocation
    b, err := req.Marshal()
    if err != nil {
        panic(err)
    }
    return string(b)
}
```

Although this costs more space to save the key, but this is loss-less and free of conflicts. There might be conflict using hash32, and we have to check the conflict.